From e45d98d095be77a0aee6cdc03e682e7713a097ff Mon Sep 17 00:00:00 2001
From: Li Hua <hucool.lihua@huawei.com>
Date: Tue, 7 Sep 2021 12:56:03 +0000
Subject: [PATCH 3320/4585] [Huawei] sched/idle: Reported an error when an
 illegal negative value is passed

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 180842

--------------------------------

Reported an error when an illegal negative value is passed. The allowed value is
0 to INT_MAX.

Signed-off-by: Li Hua <hucool.lihua@huawei.com>
Reviewed-by: Chen Hui <judy.chenhui@huawei.com>
Signed-off-by: Chen Jun <chenjun102@huawei.com>
Patchwork Links: http://patchwork.huawei.com/project/hulk5.10/list/?series=18668
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 include/linux/kernel.h |  2 +-
 kernel/sched/idle.c    | 10 +++++-----
 kernel/sysctl.c        |  6 +++---
 3 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/include/linux/kernel.h b/include/linux/kernel.h
index dace1d9dffa6..601ecfb13690 100644
--- a/include/linux/kernel.h
+++ b/include/linux/kernel.h
@@ -558,7 +558,7 @@ extern int sysctl_panic_on_stackoverflow;
 
 extern bool crash_kexec_post_notifiers;
 #ifdef CONFIG_IAS_SMART_IDLE
-extern unsigned long poll_threshold_ns;
+extern int poll_threshold_ns;
 #endif
 
 
diff --git a/kernel/sched/idle.c b/kernel/sched/idle.c
index 13460e8694bf..256d51d65066 100644
--- a/kernel/sched/idle.c
+++ b/kernel/sched/idle.c
@@ -18,7 +18,7 @@ extern char __cpuidle_text_start[], __cpuidle_text_end[];
  * Poll_threshold_ns indicates the maximum polling time before
  * entering real idle.
  */
-unsigned long poll_threshold_ns;
+int poll_threshold_ns;
 #endif
 
 /**
@@ -63,9 +63,9 @@ __setup("hlt", cpu_idle_nopoll_setup);
 #ifdef CONFIG_IAS_SMART_IDLE
 /* looping 2000 times is probably microsecond level for 2GHZ CPU*/
 #define MICRO_LEVEL_COUNT 2000
-static inline void delay_relax(unsigned long delay_max)
+static inline void delay_relax(int delay_max)
 {
-	unsigned long delay_count = 0;
+	int delay_count = 0;
 
 	delay_max = (delay_max < MICRO_LEVEL_COUNT) ? delay_max : MICRO_LEVEL_COUNT;
 	while (unlikely(!tif_need_resched()) && delay_count < delay_max) {
@@ -77,7 +77,7 @@ static inline void delay_relax(unsigned long delay_max)
 
 static inline void smart_idle_poll(void)
 {
-	unsigned long poll_duration = poll_threshold_ns;
+	int poll_duration = poll_threshold_ns;
 	ktime_t cur, stop;
 
 	if (likely(!poll_duration))
@@ -309,7 +309,7 @@ static void do_idle(void)
 {
 	int cpu = smp_processor_id();
 #ifdef CONFIG_IAS_SMART_IDLE
-	unsigned long idle_poll_flag = poll_threshold_ns;
+	int idle_poll_flag = poll_threshold_ns;
 #endif
 	/*
 	 * If the arch has a polling bit, we maintain an invariant:
diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index 334961c7783d..84c29ae6f9e7 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -1692,9 +1692,10 @@ static struct ctl_table ias_table[] = {
 	{
 		.procname	= "smart_idle_threshold",
 		.data		= &poll_threshold_ns,
-		.maxlen		= sizeof(unsigned long),
+		.maxlen		= sizeof(int),
 		.mode		= 0644,
-		.proc_handler	= proc_doulongvec_minmax,
+		.proc_handler	= proc_dointvec_minmax,
+		.extra1		= SYSCTL_ZERO,
 	},
 #endif
 
@@ -1899,7 +1900,6 @@ static struct ctl_table kern_table[] = {
 		.proc_handler	= sysctl_sched_uclamp_handler,
 	},
 #endif
-
 #ifdef CONFIG_SCHED_AUTOGROUP
 	{
 		.procname	= "sched_autogroup_enabled",
-- 
2.17.1

